src/sdam/server.ts

```diff
@@ -68,52 +80,64 @@ const stateTransition = makeStateMachine({
 
 const kMonitor = Symbol('monitor');
 
+export interface ServerOptions extends ConnectionPoolOptions, ClientMetadataOptions {
+  credentials?: MongoCredentials;
+}
+
+interface ServerPrivate {
+  /** The server description for this server */
+  description: ServerDescription;
+  /** A copy of the options used to construct this instance */
+  options?: ServerOptions;
+  /** A logger instance */
+  logger: Logger;
+  /** The current state of the Server */
+  state: string;
+  /** The topology this server is a part of */
+  topology: Topology;
+  /** A connection pool for this server */
+  pool: ConnectionPool;
+}
+
+export interface DestroyOptions {
+  force?: boolean;
+}
+
 /**
  * @fires Server#serverHeartbeatStarted
  * @fires Server#serverHeartbeatSucceeded
  * @fires Server#serverHeartbeatFailed
  */
-class Server extends EventEmitter {
-  s: any;
-  clusterTime: any;
+export class Server extends EventEmitter {
+  s: ServerPrivate;
+  clusterTime?: ClusterTime;
   ismaster?: Document;
-  [kMonitor]: any;
+  [kMonitor]: Monitor;
 
   /**
    * Create a server
-   *
-   * @param {ServerDescription} description
-   * @param {any} options
-   * @param {any} topology
    */
-  constructor(description: ServerDescription, options: any, topology: any) {
+  constructor(topology: Topology, description: ServerDescription, options?: ServerOptions) {
     super();
 
     this.s = {
-      // the server description
       description,
-      // a saved copy of the incoming options
       options,
-      // the server logger
       logger: new Logger('Server', options),
-      // the server state
       state: STATE_CLOSED,
-      credentials: options.credentials,
-      topology
+      topology,
+      pool: new ConnectionPool(
+        Object.assign({ host: description.host, port: description.port }, options)
```

Comment:
I would use a spread here to help type this down the road.

```
{ ..options, host: description.host, port: description.port }
```

Reply:
I just tested it `Object.assign` does type the return value, but we don't need it anymore in master.

Reply:
This is strange, `Object.assign` works but spreading does not with the following error:
```
Argument of type '{ host: string; port: number; } | { host: string; port: number; credentials?: MongoCredentials | undefined; maxPoolSize: number; minPoolSize: number; maxIdleTimeMS: number; ... 60 more ...; appname?: string | undefined; }' is not assignable to parameter of type 'ConnectionPoolOptions'.
  Type '{ host: string; port: number; }' is missing the following properties from type 'ConnectionPoolOptions': maxPoolSize, minPoolSize, maxIdleTimeMS, waitQueueTimeoutMS, and 8 more.ts(2345)
```

Reply:
Yeah here changing it to a spread actually shows us some type errors. The min / max pool stuff in `ConnectionPoolOptions` should have `?` then a bunch of the other options in `ConnectionOptions` might need them as well.

Reply:
`Object.assign` is not to be trusted! 👻